sqlx-core/src/query_builder.rs

```diff
@@ -0,0 +1,141 @@
+use std::fmt::Display;
+
+use crate::arguments::Arguments;
+use crate::database::{Database, HasArguments};
+use crate::encode::Encode;
+use crate::query::Query;
+use crate::types::Type;
+use either::Either;
+use std::marker::PhantomData;
+
+pub struct QueryBuilder<'a, DB>
+where
+    DB: Database,
+{
+    query: String,
+    arguments: Option<<DB as HasArguments<'a>>::Arguments>,
+    variable_count: u16,
+}
+
+impl<'a, DB: Database> QueryBuilder<'a, DB>
+where
+    DB: Database,
+{
+    pub fn new(init: impl Into<String>) -> Self
+    where
+        <DB as HasArguments<'a>>::Arguments: Default,
+    {
+        QueryBuilder {
+            query: init.into(),
+            arguments: Some(Default::default()),
+            variable_count: 0,
+        }
+    }
+
+    pub fn push(&mut self, sql: impl Display) -> &mut Self {
+        self.query.push_str(&format!(" {}", sql));
+
+        self
+    }
+
+    pub fn push_bind<A>(&mut self, value: A) -> &mut Self
+    where
+        A: 'a + Encode<'a, DB> + Send + Type<DB>,
+    {
+        match self.arguments {
+            Some(ref mut arguments) => {
+                arguments.add(value);
+                self.variable_count += 1;
+                self.query
+                    .push_str(&arguments.place_holder(self.variable_count));
+            }
+            None => panic!("Arguments taken already"),
```

Comment:
I could potentially see someone wanting to reuse the same `QueryBuilder` multiple times, maybe to amortize the allocation of the query string.

In that case, we should provide a `.clear()` or `.reset()` (bikeshedding?) method to clear the query string and replace `arguments` with `Default::default()`. We could use `self.arguments.is_none()` in `.push()` as a sanity check to remind the user to call `.reset()` after `.build()` if they want to build a new query with the same builder.